framework/src/dirackernels/ReporterPointSource.C

```diff
@@ -86,31 +102,46 @@ ReporterPointSource::addPoints()
   errorCheck("weight_name", _weight.size());
   errorCheck("point_name", _point.size());
 
-  if (isParamValid("point_name"))
+  if (_read_in_points)
   {
-    for (const auto i : index_range(_values))
-    {
-      _point_to_index[_point[i]] = i;
-      addPoint(_point[i], i);
-    }
+    for (const auto & i : index_range(_point))
+      fillPoint(_point[i], i);
   }
   else
   {
+    std::vector<Point> points(_coordx.size());
     for (const auto i : index_range(_values))
     {
-      const Point pt(_coordx[i], _coordy[i], _coordz[i]);
-      _point_to_index[pt] = i;
-      addPoint(pt, i);
+      const Point point = Point(_coordx[i], _coordy[i], _coordz[i]);
+      fillPoint(point, i);
     }
   }
 }
 
 Real
 ReporterPointSource::computeQpResidual()
 {
-  //  This is negative because it's a forcing function that has been brought over to the left side
-  std::size_t index = _point_to_index[_current_point];
-  return -_test[_i][_qp] * _values[index] * _weight[index];
+  // This is negative because it's a forcing function that has been brought over to the left side
+  return -_test[_i][_qp] * _point_to_weightedValue.at(_current_point);
+}
+
+void
+ReporterPointSource::fillPoint(const Point & point, const dof_id_type id)
+{
+  auto it = _point_to_weightedValue.find(point);
+  if (it == _point_to_weightedValue.end())
+  {
+    addPoint(point, id);
+    it = _point_to_weightedValue.emplace(point, 0).first;
+  }
+  else if (!_combine_duplicates)
+    mooseError("combine_duplicates must be true if reporter has duplicate points.  Found "
+               "duplicate point (",
+               point,
+               ").");
+
+  auto & value = it->second;
+  value += _values[id] * _weight[id];
```

Comment:
there's a fundamental change here.
you used to use the current value of the reporter when adding to the residual.
now they are cached in a map and never updated

Reply:
Every time `addPoints` is called for a Dirac kernel, which is every residual evaluation(IIRC), then `_point_to_weightedValue` is cleared and recalculated. Since the x,y,z point locations can change arbitrarily, it has to be done this way in case the point location is changed, as well as the values.

Reply:
I made this change. we do clear one each residual call https://github.com/idaholab/moose/blob/e4e6c092437ebbd0e0e6830b6e4b87ce750ce323/framework/src/systems/NonlinearSystemBase.C#L3190

Reply:
Ok that works then.